README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# pathviewr

<!-- badges: start -->

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![R build
status](https://github.com/ropensci/pathviewr/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/pathviewr/actions)
[![Codecov test
coverage](https://codecov.io/gh/ropensci/pathviewr/graph/badge.svg)](https://app.codecov.io/gh/ropensci/pathviewr?branch=master)
[![](https://badges.ropensci.org/409_status.svg)](https://github.com/ropensci/software-review/issues/409)  
[![DOI](https://zenodo.org/badge/268906628.svg)](https://zenodo.org/badge/latestdoi/268906628)
[![CRAN
status](https://www.r-pkg.org/badges/version/pathviewr)](https://CRAN.R-project.org/package=pathviewr)

<!-- badges: end -->

`pathviewr` offers tools to import, clean, and visualize movement data,
particularly from motion capture systems such as [Optitrack’s
Motive](https://optitrack.com/software/motive/), the [Straw Lab’s
Flydra](https://github.com/strawlab/flydra), or other sources. We
provide functions to remove artifacts, standardize tunnel position and
tunnel axes, select a region of interest, isolate specific trajectories,
fill gaps in trajectory data, and calculate 3D and per-axis velocity.
For experiments of visual guidance, we also provide functions that use
subject position to estimate perception of visual stimuli.

## Installation

You can install `pathviewr` from CRAN via:

``` r
install.packages("pathviewr")
```

Or to get the latest (developmental) version through GitHub, use:

``` r
devtools::install_github("ropensci/pathviewr")
```

## Example

#### Data import and cleaning via `pathviewr`

We’ll also load two `tidyverse` packages for wrangling & plotting in
this readme.

``` r
library(pathviewr)
library(ggplot2)
library(magrittr)
```

We will import and clean a sample data set from `.csv` files exported by
Optitrack’s [Motive](https://optitrack.com/software/motive/) software.
For examples of how to import and clean other types of data, [see the
Basics of data import and cleaning
vignette](https://docs.ropensci.org/pathviewr/articles/data-import-cleaning.html).

``` r
## Import the Motive example data included in 
## the package

motive_data <-
  read_motive_csv(
    system.file("extdata", "pathviewr_motive_example_data.csv",
                package = 'pathviewr')
  )
```

Several functions to clean and wrangle data are available, and we have a
suggested pipeline for how these steps should be handled. For this
example, we will use one of two “all-in-one” functions: `clean_viewr()`.
[See the Basics of data import and cleaning
vignette](https://docs.ropensci.org/pathviewr/articles/data-import-cleaning.html)
for the full pipeline and the other “all-in-one” function.

``` r
motive_allinone <-
  motive_data %>%
  clean_viewr(
    relabel_viewr_axes = TRUE,
    gather_tunnel_data = TRUE,
    trim_tunnel_outliers = TRUE,
    standardization_option = "rotate_tunnel",
    select_x_percent = TRUE,
    desired_percent = 50,
    rename_viewr_characters = FALSE,
    separate_trajectories = TRUE,
    max_frame_gap = "autodetect",
    get_full_trajectories = TRUE,
    span = 0.95
  )
#> autodetect is an experimental feature -- please report issues.

## Quick plot
## Colors correspond to unique trajectories (file_sub_traj)
motive_allinone %>%
  ggplot(aes(x = position_length, y = position_width, 
             fill = file_sub_traj)) +
  geom_point(pch = 21) +
  coord_fixed() +
  theme_classic() +
  theme(
    legend.position = "none"
  )
```

<img src="man/figures/README-all_in_one-1.png" width="100%" />

To get a sense of what we’ve done, compare the data before and after it
has passed through the pipeline.

``` r
## Check out the data's structure before cleaning and wrangling:
str(motive_data)
#> tibble [934 × 26] (S3: tbl_df/tbl/data.frame)
#>  $ frame                     : int [1:934] 72210 72211 72212 72213 72214 72215 72216 72217 72218 72219 ...
#>  $ time_sec                  : num [1:934] 722 722 722 722 722 ...
#>  $ device02_rotation_x       : num [1:934] 0.1346 0.0819 0.2106 0.1961 0.1305 ...
#>  $ device02_rotation_y       : num [1:934] -0.977 -0.978 -0.973 -0.972 -0.975 ...
#>  $ device02_rotation_z       : num [1:934] -0.1117 -0.0991 -0.0939 -0.1275 -0.1213 ...
#>  $ device02_rotation_w       : num [1:934] 0.1215 0.1654 0.0311 0.0351 0.1315 ...
#>  $ device02_position_x       : num [1:934] 0.142 0.137 0.125 0.118 0.113 ...
#>  $ device02_position_y       : num [1:934] 0.16 0.164 0.166 0.168 0.173 ...
#>  $ device02_position_z       : num [1:934] 2 1.97 1.95 1.92 1.89 ...
#>  $ device02_mean_marker_error: num [1:934] 0.000113 0.000105 0.000115 0.000202 0.000106 0.000095 0.000114 0.000117 0.000121 0.000131 ...
#>  $ device03_rotation_x       : num [1:934] 0.107 0.111 0.109 0.109 0.108 ...
#>  $ device03_rotation_y       : num [1:934] -0.295 -0.295 -0.295 -0.295 -0.295 ...
#>  $ device03_rotation_z       : num [1:934] -0.088 -0.0866 -0.0853 -0.0853 -0.0879 ...
#>  $ device03_rotation_w       : num [1:934] 0.945 0.945 0.945 0.945 0.945 ...
#>  $ device03_position_x       : num [1:934] 0.222 0.222 0.222 0.222 0.222 ...
#>  $ device03_position_y       : num [1:934] 0.245 0.245 0.245 0.245 0.245 ...
#>  $ device03_position_z       : num [1:934] 0.0597 0.0597 0.0598 0.0598 0.0598 ...
#>  $ device03_mean_marker_error: num [1:934] 0.000166 0.000172 0.000164 0.000163 0.000162 0.000162 0.000169 0.00017 0.00017 0.000213 ...
#>  $ device05_rotation_x       : num [1:934] 0.00672 0.00714 0.00709 0.00742 0.00826 ...
#>  $ device05_rotation_y       : num [1:934] 0.944 0.944 0.944 0.944 0.944 ...
#>  $ device05_rotation_z       : num [1:934] -0.117 -0.116 -0.118 -0.118 -0.117 ...
#>  $ device05_rotation_w       : num [1:934] 0.308 0.308 0.309 0.31 0.308 ...
#>  $ device05_position_x       : num [1:934] 0.173 0.173 0.173 0.173 0.173 ...
#>  $ device05_position_y       : num [1:934] 0.243 0.243 0.243 0.243 0.243 ...
#>  $ device05_position_z       : num [1:934] 2.66 2.66 2.66 2.66 2.66 ...
#>  $ device05_mean_marker_error: num [1:934] 0.000241 0.000247 0.000255 0.000244 0.00023 0.000226 0.000231 0.000236 0.000242 0.000263 ...
#>  - attr(*, ".internal.selfref")=<externalptr> 
#>  - attr(*, "pathviewr_steps")= chr "viewr"
#>  - attr(*, "file_id")= chr "pathviewr_motive_example_data.csv"
#>  - attr(*, "file_mtime")= POSIXct[1:1], format: "2023-03-07 11:12:49"
#>  - attr(*, "frame_rate")= num 100
#>  - attr(*, "header")='data.frame':   11 obs. of  2 variables:
#>   ..$ metadata: chr [1:11] "Format Version" "Take Name" "Take Notes" "Capture Frame Rate" ...
#>   ..$ value   : chr [1:11] "1.23" "sept-18_mixed-group_16-30" "" "100.000000" ...
#>  - attr(*, "Motive_IDs")= chr [1:24] "\"9E207518D8A311E969D7AB6B1FACE49B\"" "\"9E207518D8A311E969D7AB6B1FACE49B\"" "\"9E207518D8A311E969D7AB6B1FACE49B\"" "\"9E207518D8A311E969D7AB6B1FACE49B\"" ...
#>  - attr(*, "subject_names_full")= chr [1:24] "device02" "device02" "device02" "device02" ...
#>  - attr(*, "subject_names_simple")= chr [1:3] "device02" "device03" "device05"
#>  - attr(*, "data_names")= chr [1:26] "frame" "time_sec" "device02_rotation_x" "device02_rotation_y" ...
#>  - attr(*, "data_types_full")= chr [1:24] "Rigid Body" "Rigid Body" "Rigid Body" "Rigid Body" ...
#>  - attr(*, "data_types_simple")= chr "Rigid Body"
#>  - attr(*, "d1")= chr [1:26] "" "" "Rotation" "Rotation" ...
#>  - attr(*, "d2")= chr [1:26] "Frame" "Time (Seconds)" "X" "Y" ...
#>  - attr(*, "import_method")= chr "motive"

## Check out the data's structure after cleaning and wrangling:
str(motive_allinone)
#> tibble [449 × 24] (S3: tbl_df/tbl/data.frame)
#>  $ frame            : int [1:449] 72213 72214 72215 72216 72217 72218 72219 72220 72221 72222 ...
#>  $ time_sec         : num [1:449] 722 722 722 722 722 ...
#>  $ subject          : chr [1:449] "device02" "device02" "device02" "device02" ...
#>  $ position_length  : num [1:449] 0.647 0.62 0.593 0.567 0.541 ...
#>  $ position_width   : num [1:449] -0.112 -0.116 -0.122 -0.134 -0.141 ...
#>  $ position_height  : num [1:449] -0.0371 -0.0324 -0.0273 -0.0235 -0.0209 ...
#>  $ rotation_length  : num [1:449] -0.128 -0.121 -0.105 -0.106 -0.149 ...
#>  $ rotation_width   : num [1:449] 0.1961 0.1305 0.0935 0.1798 0.164 ...
#>  $ rotation_height  : num [1:449] -0.972 -0.975 -0.975 -0.975 -0.972 ...
#>  $ rotation_real    : num [1:449] 0.0351 0.1315 0.1734 0.0807 0.0824 ...
#>  $ mean_marker_error: num [1:449] 0.000202 0.000106 0.000095 0.000114 0.000117 0.000121 0.000131 0.00014 0.000113 0.000114 ...
#>  $ velocity         : num [1:449] 2.73 2.78 2.84 2.85 2.68 ...
#>  $ length_inst_vel  : num [1:449] -2.65 -2.72 -2.74 -2.58 -2.56 ...
#>  $ width_inst_vel   : num [1:449] -0.642 -0.387 -0.58 -1.139 -0.75 ...
#>  $ height_inst_vel  : num [1:449] 0.184 0.475 0.508 0.379 0.258 ...
#>  $ traj_id          : int [1:449] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ file_sub_traj    : chr [1:449] "pathviewr_motive_example_data.csv_device02_0" "pathviewr_motive_example_data.csv_device02_0" "pathviewr_motive_example_data.csv_device02_0" "pathviewr_motive_example_data.csv_device02_0" ...
#>  $ traj_length      : int [1:449] 63 63 63 63 63 63 63 63 63 63 ...
#>  $ start_length     : num [1:449] 0.647 0.647 0.647 0.647 0.647 ...
#>  $ end_length       : num [1:449] -0.656 -0.656 -0.656 -0.656 -0.656 ...
#>  $ length_diff      : num [1:449] 1.3 1.3 1.3 1.3 1.3 ...
#>  $ start_length_sign: num [1:449] 1 1 1 1 1 1 1 1 1 1 ...
#>  $ end_length_sign  : num [1:449] -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 ...
#>  $ direction        : chr [1:449] "leftwards" "leftwards" "leftwards" "leftwards" ...
#>  - attr(*, "file_id")= chr "pathviewr_motive_example_data.csv"
#>  - attr(*, "file_mtime")= POSIXct[1:1], format: "2023-03-07 11:12:49"
#>  - attr(*, "frame_rate")= num 100
#>  - attr(*, "header")='data.frame':   11 obs. of  2 variables:
#>   ..$ metadata: chr [1:11] "Format Version" "Take Name" "Take Notes" "Capture Frame Rate" ...
#>   ..$ value   : chr [1:11] "1.23" "sept-18_mixed-group_16-30" "" "100.000000" ...
#>  - attr(*, "Motive_IDs")= chr [1:24] "\"9E207518D8A311E969D7AB6B1FACE49B\"" "\"9E207518D8A311E969D7AB6B1FACE49B\"" "\"9E207518D8A311E969D7AB6B1FACE49B\"" "\"9E207518D8A311E969D7AB6B1FACE49B\"" ...
#>  - attr(*, "subject_names_full")= chr [1:24] "device02" "device02" "device02" "device02" ...
#>  - attr(*, "subject_names_simple")= chr [1:3] "device02" "device03" "device05"
#>  - attr(*, "data_names")= chr [1:26] "frame" "time_sec" "device02_rotation_x" "device02_rotation_y" ...
#>  - attr(*, "data_types_full")= chr [1:24] "Rigid Body" "Rigid Body" "Rigid Body" "Rigid Body" ...

NEWS.md
# pathviewr 1.1.7
* Remove dependency on `anomalize` package by electing to not export 
`remove_vel_anomalies()` as the package is loaded.

# pathviewr 1.1.6
* Bug fixes to `remove_duplicate_frames()` to ensure data are not accidentally
deleted

# pathviewr 1.1.5
* Bug fixes to `plot_by_subject()` and other plotting functions to ensure 
compliance with ggplot2 v.3.4.0

# pathviewr 1.1.4
* `gather_tunnel_data()` now checks to see if mean marker error is present in
the data and skips such columns if they are absent

# pathviewr 1.1.3
* `get_2d_angle()` now accepts numeric vectors for each argument and will 
return a vector of angles of equal length

# pathviewr 1.1.2
* `viewr` attribute requirements have been largely relaxed. This is intended
to provide better flexibility of using functions at arbitrary points within the
"data analysis pipeline". 
* Users should be advised that since some safeguards have been removed, objects
passed through most `pathviewr` functions are not checked as thoroughly for all
formatting requirements.

# pathviewr 1.1.1
* Remove viewr attribute requirements. Work in progress

# pathviewr 1.1.0
* New data cleaning functions added: set_traj_frametime(),
get_traj_velocities(), clean_by_span(), remove_duplicate_frames(), and
remove_vel_anomalies()
* These new functions have not been thoroughly vetted nor have unit tests
been written for them -- please use with caution and report issues.

# pathviewr 1.0.0
* Package has been accepted by rOpenSci and is now hosted on ropensci/pathviewr
* No changes to code since v0.9.5

# pathviewr 0.9.5
* Package has been updated to incorporate feedback from rOpenSci reviewers
@asbonnetlebrun and @marcosci, along with editor @maelle. See here for more
details: ropensci/software-review#409

# pathviewr 0.9.4
* We are targeting a submission to rOpenSci in the near future (hopefully
today). This version of pathviewr has been prepped according to rOpenSci's
"Packages: Development, Maintenance, and Peer Review" guide.


R/plotting_functions.R
## Part of the pathviewr package
## Last updated: 2020-09-16 VBB & MSA


########################### visualize_frame_gap_choice #########################

#' Visualize the consequence of using various max_frame_gap values
#'
#' Run separate_trajectories() with many different frame gaps to help determine
#' what value to use
#'
#' @param obj_name The input viewr object; a tibble or data.frame with attribute
#'   \code{pathviewr_steps} that includes \code{"viewr"}
#' @param loops How many total frame gap entries to consider. Each loop will
#'   increase the \code{max_fram_gap} argument in \code{separate_trajectories}
#'   by 1.
#' @param ... Additional arguments
#'
#' @details The input viewr object (\code{obj_name}) should likely be an object
#'   that has passed through the \code{select_x_percent()} step.
#'
#' @return A plot and a tibble, each of which shows the total number of
#'   trajectories that result from using the specified range of
#'   \code{max_frame_gap} values.
#'
#' @author Melissa S. Armstrong and Vikram B. Baliga
#'
#' @family data cleaning functions
#' @family plotting functions
#' @family functions that define or clean trajectories
#'
#' @importFrom graphics plot
#'
#' @export
#'
#' @examples
#' library(pathviewr)
#'
#' ## Import the example Motive data included in the package
#' motive_data <-
#'   read_motive_csv(system.file("extdata", "pathviewr_motive_example_data.csv",
#'                              package = 'pathviewr'))
#'
#' motive_selected <-
#'   motive_data %>%
#'   relabel_viewr_axes() %>%
#'   gather_tunnel_data() %>%
#'   trim_tunnel_outliers() %>%
#'   rotate_tunnel() %>%
#'   get_velocity() %>%
#'   select_x_percent(desired_percent = 50)
#'
#' visualize_frame_gap_choice(motive_selected, loops = 10)

visualize_frame_gap_choice <- function(obj_name,
                                       loops = 20,
                                       ...){

  ## Check that it's a viewr object
  # if (!any(attr(obj_name,"pathviewr_steps") == "viewr")) {
  #   stop("This doesn't seem to be a viewr object")
  # }

  # make a bunch of empty vectors to dump info
  mfg <- vector("list", loops)
  cts <- vector("list", loops)
  trajectory_count <- vector(mode = "double", loops)
  frame_gap_allowed <- vector(mode = "double", loops)

  # loop through user defined number of max frame gap values
  i <- 1
  while (i < loops + 1) {
    mfg[[i]] <- quick_separate_trajectories(obj_name, max_frame_gap = i)
    cts[[i]] <- dplyr::count(mfg[[i]], traj_id)
    trajectory_count[i] <- nrow(cts[[i]])
    frame_gap_allowed[i] <- i
    i <- i +1
  }

  ## Collect the info on max frame gaps allowed vs. trajectory counts
  mfg_tib <- tibble::tibble(frame_gap_allowed,
                            trajectory_count)

  ## Find the curve elbow point via `find_curve_elbow()`
  max_fg <- find_curve_elbow(mfg_tib,
                             export_type = "row_num",
                             plot_curve = FALSE)

  ## Paste filename into export tibble
  obj_name_arg <- deparse(substitute(obj_name))
  mfg_tib$file_id <- as.character(obj_name_arg)

  mfg_plot <- plot(mfg_tib$frame_gap_allowed,
                   mfg_tib$trajectory_count); graphics::abline(v = max_fg)

  return(list(mfg_tib, mfg_plot))

}


########################### plot_viewr_trajectories ############################

#' Plot each trajectory within a viewr object
#'
#' @param obj_name A viewr object (a tibble or data.frame with attribute
#'   \code{pathviewr_steps} that includes \code{"viewr"}) that has been passed
#'   through \code{separate_trajectories()} or \code{get_full_trajectories()}.
#' @param plot_axes Which position axes should be plotted? A character vector
#'   including exactly two of the following choices must be supplied:
#'   \code{length}, \code{width}, \code{height}. Default is c("length",
#'   "width").
#' @param multi_plot Should separate plots (one per trajectory) be created or
#'   should one multi-plot grid be generated. Defaults to FALSE, which produces
#'   separate plots.
#'
#' @return A (base-R) series of plots or single plot (if \code{multi_plot =
#'   TRUE}) that depict each trajectory along the chosen axes.
#'
#' @export
#'
#' @importFrom graphics plot
#'
#' @author Vikram B. Baliga
#'
#' @family plotting functions
#'
#' @examples
#' library(pathviewr)
#'
#' ## Import the example Motive data included in the package
#' motive_data <-
#'   read_motive_csv(system.file("extdata", "pathviewr_motive_example_data.csv",
#'                              package = 'pathviewr'))
#'
#' motive_full <-
#'   motive_data %>%
#'   clean_viewr(desired_percent = 50,
#'               max_frame_gap = "autodetect",
#'               span = 0.95)
#'
#' plot_viewr_trajectories(motive_full, multi_plot = FALSE)
#' plot_viewr_trajectories(motive_full, multi_plot = TRUE)

plot_viewr_trajectories <- function(obj_name,
                                    plot_axes = c("length", "width"),
                                    multi_plot = FALSE){

  ## Collect names of trajectories
  obj_name_trajs <- unique(obj_name$file_sub_traj)
  ## For multi-plotting, find out what the square root of the total number
  ## of trajectories is. This will be used to set par(mfrow()).
  sqrt_traj_count <- ceiling(sqrt(length(obj_name_trajs)))

  oldpar <- par(no.readonly = TRUE)
  on.exit(par(oldpar))

  if (plot_axes[1] == "length"){
    if (plot_axes[2] == "width") {
      if (multi_plot == TRUE) {
        oldpar <- par(no.readonly = TRUE)
        on.exit(par(oldpar))
        par(mar = c(0, 0, 0, 0))
        par(mfrow = c(sqrt_traj_count, sqrt_traj_count))
      }
      for (i in seq_len(length(obj_name_trajs))){
        tmp <- obj_name %>% dplyr::filter(file_sub_traj == obj_name_trajs[i])
        plot(tmp$position_length,
             tmp$position_width,
             asp = 1,
             ## add a title that indicates sub_traj
             main = obj_name_trajs[i],
             ## add sub-title to indicate i
             sub = paste0("trajectory #",i),
             ## keep the same dimensions across all plots:
             xlim = c(min(obj_name$position_length),
                      max(obj_name$position_length)),
             ylim = c(min(obj_name$position_width),
                      max(obj_name$position_width))
        )
      }
    }
    if (plot_axes[2] == "height") {
      if (multi_plot == TRUE) {
        oldpar <- par(no.readonly = TRUE)
        on.exit(par(oldpar))
        par(mar = c(0, 0, 0, 0))
        par(mfrow = c(sqrt_traj_count, sqrt_traj_count))
      }
      for (i in seq_len(length(obj_name_trajs))){
        tmp <- obj_name %>% dplyr::filter(file_sub_traj == obj_name_trajs[i])
        plot(tmp$position_length,
             tmp$position_height,
             asp = 1,
             ## add a title that indicates sub_traj
             main = obj_name_trajs[i],
             ## add sub-title to indicate i
             sub = paste0("trajectory #",i),
             ## keep the same dimensions across all plots:
             xlim = c(min(obj_name$position_length),
                      max(obj_name$position_length)),
             ylim = c(min(obj_name$position_height),
                      max(obj_name$position_height))
        )
      }
    }
  }

  if (plot_axes[1] == "width"){
    if (plot_axes[2] == "length") {
      if (multi_plot == TRUE) {
        oldpar <- par(no.readonly = TRUE)
        on.exit(par(oldpar))
        par(mar = c(0, 0, 0, 0))
        par(mfrow = c(sqrt_traj_count, sqrt_traj_count))
      }
      for (i in seq_len(length(obj_name_trajs))){
        tmp <- obj_name %>% dplyr::filter(file_sub_traj == obj_name_trajs[i])
        plot(tmp$position_width,
             tmp$position_length,
             asp = 1,
             ## add a title that indicates sub_traj
             main = obj_name_trajs[i],
             ## add sub-title to indicate i
             sub = paste0("trajectory #",i),
             ## keep the same dimensions across all plots:
             xlim = c(min(obj_name$position_width),
                      max(obj_name$position_width)),
             ylim = c(min(obj_name$position_length),
                      max(obj_name$position_length))
        )
      }
    }
    if (plot_axes[2] == "height") {
      if (multi_plot == TRUE) {
        oldpar <- par(no.readonly = TRUE)
        on.exit(par(oldpar))
        par(mar = c(0, 0, 0, 0))
        par(mfrow = c(sqrt_traj_count, sqrt_traj_count))
      }
      for (i in seq_len(length(obj_name_trajs))){
        tmp <- obj_name %>% dplyr::filter(file_sub_traj == obj_name_trajs[i])
        plot(tmp$position_width,
             tmp$position_height,
             asp = 1,
             ## add a title that indicates sub_traj
             main = obj_name_trajs[i],
             ## add sub-title to indicate i
             sub = paste0("trajectory #",i),
             ## keep the same dimensions across all plots:
             xlim = c(min(obj_name$position_width),
                      max(obj_name$position_width)),
             ylim = c(min(obj_name$position_height),
                      max(obj_name$position_height))
        )
      }
    }
  }

  if (plot_axes[1] == "height"){
    if (plot_axes[2] == "length") {
      oldpar <- par(no.readonly = TRUE)
      on.exit(par(oldpar))
      if (multi_plot == TRUE) {
        par(mar = c(0, 0, 0, 0))
        par(mfrow = c(sqrt_traj_count, sqrt_traj_count))
      }
      for (i in seq_len(length(obj_name_trajs))){
        tmp <- obj_name %>% dplyr::filter(file_sub_traj == obj_name_trajs[i])
        plot(tmp$position_height,
             tmp$position_length,
             asp = 1,
             ## add a title that indicates sub_traj
             main = obj_name_trajs[i],
             ## add sub-title to indicate i
             sub = paste0("trajectory #",i),
             ## keep the same dimensions across all plots:
             xlim = c(min(obj_name$position_height),
                      max(obj_name$position_height)),
             ylim = c(min(obj_name$position_length),
                      max(obj_name$position_length))
        )
      }
    }
    if (plot_axes[2] == "width") {
      if (multi_plot == TRUE) {
        oldpar <- par(no.readonly = TRUE)
        on.exit(par(oldpar))
        par(mar = c(0, 0, 0, 0))
        par(mfrow = c(sqrt_traj_count, sqrt_traj_count))
      }
      for (i in seq_len(length(obj_name_trajs))){
        tmp <- obj_name %>% dplyr::filter(file_sub_traj == obj_name_trajs[i])
        plot(tmp$position_height,
             tmp$position_width,
             asp = 1,
             ## add a title that indicates sub_traj
             main = obj_name_trajs[i],
             ## add sub-title to indicate i
             sub = paste0("trajectory #",i),
             ## keep the same dimensions across all plots:
             xlim = c(min(obj_name$position_height),
                      max(obj_name$position_height)),
             ylim = c(min(obj_name$position_width),
                      max(obj_name$position_width))
        )
      }
    }
  }

  ## reset back to default
  on.exit(par(oldpar))

}

####################### plot_by_subject ##########################

#' Plot trajectories and density plots of position by subject
#'
#' Plots all trajectories and generates density plots of position by subject
#' from elevation and bird's eye views.
#'
#' @param obj_name A viewr object (a tibble or data.frame with attribute
#'   \code{pathviewr_steps} that includes \code{"viewr"}) that has been passed
#'   through \code{separate_trajectories()} or \code{get_full_trajectories()}.
#' @param col_by_treat If multiple treatments or sessions, color data per
#'   treatment or session. Treatments must be levels in a column named
#'   \code{treatment}.
#' @param ... Additional arguments passed to/from other pathviewr functions.
#'
#' @details  The input viewr object should have passed through
#'   \code{separate_trajectories()} or \code{get_full_trajectories()}.
#'   Optionally, treatments should have been added as levels in a column named
#'   \code{treatment}. Two plots will be produced, one from a "bird's eye view"
#'   of width against length and one from an "elevation view" of height against
#'   length. All trajectories will be plotted on a per subject basis, along with
#'   density plots of width or height depending on the view.
#'   \code{col_by_treat = TRUE}, data will be plotted by color according to
#'   treatment in both the trajectory plots and the density plots.
#'
#' @return A "bird's eye view" plot and an "elevation view" plot, made via
#'   ggplot2.
#'
#' @importFrom graphics plot
#'
#' @export
#' @author Melissa S. Armstrong
#' @family plotting functions
#'
#' @examples
#' library(pathviewr)
#' library(ggplot2)
#' library(magrittr)
#'
#' if (interactive()) {
#'   ## Import the example Motive data included in the package
#'   motive_data <-
#'     read_motive_csv(system.file("extdata",
#'                                 "pathviewr_motive_example_data.csv",
#'                                 package = 'pathviewr'))
#'
#'   ## Clean, isolate, and label trajectories
#'   motive_full <-
#'     motive_data %>%
#'     clean_viewr(desired_percent = 50,
#'                 max_frame_gap = "autodetect",
#'                 span = 0.95)
#'
#'   ## Plot all trajectories by subject
#'   motive_full %>%
#'     plot_by_subject()
#'
#'   ## Add treatment information
#'   motive_full$treatment <- c(rep("latA", 100), rep("latB", 100),
#'                              rep("latA", 100), rep("latB", 149))
#'
#'   ## Plot all trajectories by subject, color by treatment
#'   motive_full %>%
#'     plot_by_subject(col_by_treat = TRUE)
#' }


plot_by_subject <- function(obj_name,
                            col_by_treat = FALSE,
                            ...) {

  ## Check that it's a viewr object
  # if (!any(attr(obj_name, "pathviewr_steps") == "viewr")) {
  #   stop("This doesn't seem to be a viewr object")
  # }

  #set axes limits based on data
  height_limits <- c(max(abs(range(
    obj_name$position_height
  ))) * -1,
  max(abs(range(
    obj_name$position_height
  ))))
  width_limits <- c(max(abs(range(
    obj_name$position_width
  ))) * -1,
  max(abs(range(
    obj_name$position_width
  ))))

  if (col_by_treat == FALSE) {
    #for top view
    top_view <- obj_name %>%
      dplyr::group_by(subject) %>%
      tidyr::nest() %>%
      dplyr::mutate(
        paths = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(
            position_length,
            position_width
          )) +
            ggplot2::geom_point(alpha = .1, show.legend = FALSE) +
            ggplot2::ylim(width_limits) +
            ggplot2::geom_hline(yintercept = 0, linetype = "dotted") +
            ggplot2::coord_fixed(ratio = 1)
        ),
        hist = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(x = position_width)) +
            ggplot2::geom_density(
              alpha = .5,
              position = "identity",
              show.legend = FALSE
            ) +
            ggplot2::xlim(width_limits) +
            ggplot2::geom_vline(xintercept = 0, linetype = "dotted") +
            ggplot2::coord_flip()
        )
      )

    #all of them together:
    top_all_plots <- top_view %>%
      dplyr::select(subject, paths, hist) %>%
      tidyr::gather("plot_type", "allplots", 2:3)

    p1 <- NULL
    cowA <- cowplot::plot_grid(p1, labels = "")
    cowB <- cowplot::plot_grid(
      plotlist = top_all_plots[[3]],
      labels = top_all_plots$subject,
      label_size = 8,
      label_y = 1.1
    )

    birdseye_view <-
      cowplot::plot_grid(cowA, cowB, ncol = 1, rel_heights = c(0.05, 1))

    #for elev view
    elev_view <- obj_name %>%
      dplyr::group_by(subject) %>%
      tidyr::nest() %>%
      dplyr::mutate(
        paths = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(
            position_length, position_height
          )) +
            ggplot2::geom_point(alpha = .1, show.legend = FALSE) +
            ggplot2::ylim(height_limits) +
            ggplot2::geom_hline(yintercept = 0, linetype = "dotted") +
            ggplot2::coord_fixed(ratio = 1)
        ),
        hist = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(position_height)) +
            ggplot2::geom_density(
              alpha = .5,
              position = "identity",
              show.legend = FALSE
            ) +
            ggplot2::xlim(height_limits) +
            ggplot2::geom_vline(xintercept = 0, linetype = "dotted") +
            ggplot2::coord_flip()
        )
      )

    #all of them together:
    elev_all_plots <- elev_view %>%
      dplyr::select(subject, paths, hist) %>%
      tidyr::gather("plot_type", "allplots", 2:3)

    cowC <- cowplot::plot_grid(
      plotlist = elev_all_plots[[3]],
      labels = elev_all_plots$subject,
      label_size = 8,
      label_y = 1.1
    )

    side_view <-
      cowplot::plot_grid(cowA, cowC, ncol = 1, rel_heights = c(0.05, 1))

    return(list(birdseye_view, side_view))
  }

  if (col_by_treat == TRUE) {
    #for top view (change in lateral position)
    top_view <- obj_name %>%
      dplyr::group_by(subject) %>%
      tidyr::nest() %>%
      dplyr::mutate(
        paths = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(
            position_length,
            position_width,
            colour = treatment
          )) +
            ggplot2::geom_point(alpha = .1, show.legend = FALSE) +
            ggplot2::ylim(width_limits) +
            ggplot2::geom_hline(yintercept = 0, linetype = "dotted") +
            ggplot2::coord_fixed(ratio = 1)
        ),
        hist = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(x = position_width,
                                   fill = treatment)) +
            ggplot2::geom_density(
              alpha = .5,
              position = "identity",
              show.legend = FALSE
            ) +
            ggplot2::xlim(width_limits) +
            ggplot2::geom_vline(xintercept = 0, linetype = "dotted") +
            ggplot2::coord_flip()
        )
      )

    #all of them together:
    top_all_plots <- top_view %>%
      dplyr::select(subject, paths, hist) %>%
      tidyr::gather("plot_type", "allplots", 2:3)

    p1 <- NULL
    cowA <- cowplot::plot_grid(p1, labels = "")
    cowB <- cowplot::plot_grid(
      plotlist = top_all_plots[[3]],
      labels = top_all_plots$subject,
      label_size = 8,
      label_y = 1.1
    )

    birdseye_view <-
      cowplot::plot_grid(cowA, cowB, ncol = 1, rel_heights = c(0.05, 1))

    #for elev view (change in height):
    elev_view <- obj_name %>%
      dplyr::group_by(subject) %>%
      tidyr::nest() %>%
      dplyr::mutate(
        paths = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(
            position_length, position_height, colour = treatment
          )) +
            ggplot2::geom_point(alpha = .1, show.legend = FALSE) +
            ggplot2::ylim(height_limits) +
            ggplot2::geom_hline(yintercept = 0, linetype = "dotted") +
            ggplot2::coord_fixed(ratio = 1)
        ),
        hist = purrr::map(
          data,
          ~ ggplot2::ggplot(., aes(position_height, fill = treatment)) +
            ggplot2::geom_density(
              alpha = .5,
              position = "identity",
              show.legend = FALSE
            ) +
            ggplot2::xlim(height_limits) +
            ggplot2::geom_vline(xintercept = 0, linetype = "dotted") +
            ggplot2::coord_flip()
        )
      )

    #all of them together:
    elev_all_plots <- elev_view %>%
      dplyr::select(subject, paths, hist) %>%
      tidyr::gather("plot_type", "allplots", 2:3)

    cowC <- cowplot::plot_grid(
      plotlist = elev_all_plots[[3]],
      labels = elev_all_plots$subject,
      label_size = 8,
      label_y = 1.1
    )

    side_view <-
      cowplot::plot_grid(cowA, cowC, ncol = 1, rel_heights = c(0.05, 1))

    return(list(birdseye_view, side_view))
  }

}
